apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,325 @@
+      "id": "_Avscjx5srFy",
+      "deleted": true
+    },
+    "_status": "synced"
+  }
+
+  [4] The schema table for 6 different types from [5]:
+  +---------------+---------------+---------------+
+  |   bookmark    | microsummary**|     query     |
+  +---------------+---------------+---------------+
+  |               | *generatorUri | *folderName   |
+  |               | *staticTitle  | *queryId      |
+  | title         | title         | title         |
+  | bmkUri        | bmkUri        | bmkUri        |
+  | description   | description   | description   |
+  | loadInSidebar | loadInSidebar | loadInSidebar |
+  | tags          | tags          | tags          |
+  | keyword       | keyword       | keyword       |
+  | parentid      | parentid      | parentid      |
+  | parentName    | parentName    | parentName    |
+  | predecessorid | predecessorid | predecessorid |
+  | type          | type          | type          |
+  +---------------+---------------+---------------+
+
+  +---------------+---------------+---------------+
+  | folder        | livemark      | separator     |
+  +---------------+---------------+---------------+
+  |               | *siteUri      | *pos          |
+  |               | *feedUri      |               |
+  | title         | title         |               |
+  | parentid      | parentid      | parentid      |
+  | parentName    | parentName    | parentName    |
+  | predecessorid | predecessorid | predecessorid |
+  | type          | type          | type          |
+  +---------------+---------------+---------------+
+
+  [*] Special property for its type.
+  [**] microsummary is OBSOLETED. Please see here:
+       https://wiki.mozilla.org/Microsummaries
+
+  [5] https://docs.services.mozilla.com/sync/objectformats.html#bookmarks
+
+**/
+  _update(remoteRecords, lastModifiedTime) {
+    var bookmarks = [];
+    for (var i = 0; i < remoteRecords.length; i++) {
+      var payload = remoteRecords[i].payload;
+      if (payload.type === 'microsummary') {
+        console.warn('microsummary is OBSOLETED ', payload);
+        continue;
+      }
+      if (!Number.isInteger(remoteRecords[i].last_modified)) {
+        console.warn('Incorrect payload::last_modified? ', payload);
+        continue;
+      }
+      if (remoteRecords[i].last_modified <= lastModifiedTime) {
+        break;
+      }
+      if (payload.deleted) {
+        bookmarks.push({
+          deleted: true,
+          fxsyncId: payload.id
+        });
+        continue;
+      }
+      var typeWithUri = ['query', 'bookmark']
+          .some(value => value === payload.type);
+      if (typeWithUri && !payload.bmkUri) {
+        console.warn('Incorrect payload? ', payload);
+        continue;
+      }
+      bookmarks.push({
+        id: payload.type === 'bookmark' ? payload.bmkUri :
+          (payload.type + '|' + payload.id),
+        url: payload.bmkUri,
+        name: payload.title,
+        type: payload.type === 'bookmark' ? 'url' : 'others',
```

Comment:
I trace the code to use 'url' type at here:
https://github.com/mozilla-b2g/gaia/blob/master/apps/system/js/app_chrome.js#L1043
I did not see any other types existing, and 'others' is for other types without a valid URL.